data/migrations/V0079__add_indexes_on_fec_fitem_sched_b_tables.sql

```diff
@@ -0,0 +1,2424 @@
+CREATE INDEX IF NOT EXISTS idx_sched_b_1975_1976_rcpt_name_text_colsc_disb_dt_sub_id ON disclosure.fec_fitem_sched_b_1975_1976 USING gin (recipient_name_text, (COALESCE(disb_dt, '9999-12-31'::date::timestamp without time zone)), sub_id);
+CREATE INDEX IF NOT EXISTS idx_sched_b_1975_1976_rpt_yr_colsc_disb_dt_sub_id ON disclosure.fec_fitem_sched_b_1975_1976 USING btree (rpt_yr, (COALESCE(disb_dt, '9999-12-31'::date::timestamp without time zone)), sub_id);
+CREATE INDEX IF NOT EXISTS idx_sched_b_1975_1976_line_num_colsc_disb_dt_sub_id ON disclosure.fec_fitem_sched_b_1975_1976 USING btree (line_num COLLATE pg_catalog."default", (COALESCE(disb_dt, '9999-12-31'::date::timestamp without time zone)), sub_id);
+CREATE INDEX IF NOT EXISTS idx_sched_b_1975_1976_disb_amt_colsc_disb_dt_sub_id ON disclosure.fec_fitem_sched_b_1975_1976 USING btree (disb_amt, (COALESCE(disb_dt, '9999-12-31'::date::timestamp without time zone)), sub_id);
+CREATE INDEX IF NOT EXISTS idx_sched_b_1975_1976_colsc_disb_dt_sub_id ON disclosure.fec_fitem_sched_b_1975_1976 USING btree ((COALESCE(disb_dt, '9999-12-31'::date::timestamp without time zone)), sub_id);
+-- disb_amt
+CREATE INDEX IF NOT EXISTS idx_sched_b_1975_1976_cln_rcpt_cmte_id_disb_amt_sub_id ON disclosure.fec_fitem_sched_b_1975_1976 USING btree (clean_recipient_cmte_id COLLATE pg_catalog."default", disb_amt, sub_id);
+CREATE INDEX IF NOT EXISTS idx_sched_b_1975_1976_rcpt_st_disb_amt_sub_id ON disclosure.fec_fitem_sched_b_1975_1976  USING btree (recipient_st COLLATE pg_catalog."default", disb_amt, sub_id);
+CREATE INDEX IF NOT EXISTS idx_sched_b_1975_1976_rcpt_city_disb_amt_sub_id ON disclosure.fec_fitem_sched_b_1975_1976 USING btree (recipient_city COLLATE pg_catalog."default", disb_amt, sub_id);
+CREATE INDEX IF NOT EXISTS idx_sched_b_1975_1976_cmte_id_disb_amt_sub_id ON disclosure.fec_fitem_sched_b_1975_1976 USING btree (cmte_id COLLATE pg_catalog."default", disb_amt, sub_id);
+CREATE INDEX IF NOT EXISTS idx_sched_b_1975_1976_disb_desc_text_disb_amt_sub_id ON disclosure.fec_fitem_sched_b_1975_1976 USING gin (disbursement_description_text, disb_amt, sub_id);
+CREATE INDEX IF NOT EXISTS idx_sched_b_1975_1976_image_num_disb_amt_sub_id ON disclosure.fec_fitem_sched_b_1975_1976 USING btree (image_num COLLATE pg_catalog."default", disb_amt, sub_id);
+CREATE INDEX IF NOT EXISTS idx_sched_b_1975_1976_rcpt_name_text_disb_amt_sub_id ON disclosure.fec_fitem_sched_b_1975_1976 USING gin (recipient_name_text, disb_amt, sub_id);
+CREATE INDEX IF NOT EXISTS idx_sched_b_1975_1976_rpt_yr_disb_amt_sub_id ON disclosure.fec_fitem_sched_b_1975_1976 USING btree (rpt_yr, disb_amt, sub_id);
+CREATE INDEX IF NOT EXISTS idx_sched_b_1975_1976_line_num_disb_amt_sub_id ON disclosure.fec_fitem_sched_b_1975_1976 USING btree (line_num COLLATE pg_catalog."default", disb_amt, sub_id);

```

Comment:
In checking this over and cross-referencing the existing indexes I noticed that we may be missing one more combination, that of the `COALESCE` `disb_dt` value paired first with the `disb_amt` and `sub_id` following.  Example:

```sql
CREATE INDEX IF NOT EXISTS idx_sched_b_xxxx_yyyy_colsc_disb_dt_disb_amt_sub_id ON disclosure.fec_fitem_sched_b_xxxx_yyyy USING btree ((COALESCE(disb_dt, '9999-12-31'::date::timestamp without time zone)), disb_amt, sub_id);
```

We have it going in the other direction with `disb_amt` coming first, but not this way, and I wonder if that's still contributing to some performance issues now that I think about it.

Reply:
In looking at this a bit more, we can filter and sort by both an amount and a date, so I think we do need this; I'm not sure why we didn't have it before, must've been an oversight!

Reply:
I had created three different indexes:
idx_sched_b_2017_2018_disb_amt_sub_id (disb_amt, sub_id);
idx_sched_b_2017_2018_colsc_disb_dt_disb_amt_sub_id ((COALESCE(disb_dt, ''9999-12-31''::date::timestamp without time zone)), disb_amt, sub_id);
idx_sched_b_2017_2018_disb_dt_amt_sub_id(disb_dt, disb_amt, sub_id);
  
I then run the following query (base on the sql from API):
explain analyze
select cmte_id, two_year_transaction_period, disb_dt
FROM disclosure.fec_fitem_sched_b 
WHERE two_year_transaction_period = 2018 
AND disb_dt >= '2018-02-01'::date
ORDER BY disb_amt DESC, sub_id DESC
LIMIT 20;

Limit  (cost=0.58..280.74 rows=20 width=40) (actual time=0.028..0.262 rows=20 loops=1)
  ->  Merge Append  (cost=0.58..48570525.81 rows=3467401 width=40) (actual time=0.027..0.251 rows=20 loops=1)
        Sort Key: fec_fitem_sched_b.disb_amt DESC, fec_fitem_sched_b.sub_id DESC
        ->  Sort  (cost=0.01..0.02 rows=1 width=94) (actual time=0.008..0.008 rows=0 loops=1)
              Sort Key: fec_fitem_sched_b.disb_amt DESC, fec_fitem_sched_b.sub_id DESC
              Sort Method: quicksort  Memory: 25kB
              ->  Seq Scan on fec_fitem_sched_b  (cost=0.00..0.00 rows=1 width=94) (actual time=0.003..0.003 rows=0 loops=1)
                    Filter: ((disb_dt >= '2018-02-01'::date) AND (two_year_transaction_period = '2018'::numeric))
        ->  Index Scan Backward using idx_sched_b_2017_2018_disb_amt_sub_id on fec_fitem_sched_b_2017_2018  (cost=0.56..48527183.27 rows=3467400 width=40) (actual time=0.017..0.235 rows=20 loops=1)
              Filter: ((disb_dt >= '2018-02-01'::date) AND (two_year_transaction_period = '2018'::numeric))
              Rows Removed by Filter: 160
Planning time: 2.922 ms
Execution time: 0.301 ms

idx_sched_b_2017_2018_disb_amt_sub_id (disb_amt, sub_id) is the index the DB optimizer used.  Based on this result, I will create this index to all the tables and re-submit the migration file again.

Reply:
updated migration to add extra indexes had been pushed to the branch.

Reply:
Wait, are we sure that's the SQL from the API?  Sorry, now I have one more question... 😆 I promise this will be the last one though. :-)

I want to verify the query being generated by the API to ensure we're basing the index creation on the right information.  The reason why I ask this is because the test queries we've generated in the past _do_ have `COALESCE` statements in them, plus a host of other things. Your investigation into what the query planner is great and definitely useful, I just want to make sure we're basing the findings on the right information, and I'm sorry I didn't provide a clearer example of what I was going for.  We can work through this tomorrow though and get it ironed out quick to wrap this up. :-)

Thanks so much for working through this with me, @fecjjeng!